logserver/src/main/java/alluxio/logserver/AlluxioLog4jSocketNode.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.log4j.PropertyConfigurator;
+import org.apache.log4j.spi.LoggerRepository;
+import org.apache.log4j.spi.LoggingEvent;
+import org.apache.log4j.spi.RootLogger;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.net.Socket;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Properties;
+
+/**
+ * Reads {@link org.apache.log4j.spi.LoggingEvent} objects from remote logging
+ * clients and writes the objects to designated log files. For each logging client,
+ * logging server creates a {@link AlluxioLog4jSocketNode} object and starts
+ * a thread serving the logging requests of the client.
+ */
+public class AlluxioLog4jSocketNode implements Runnable {
+  private static final org.slf4j.Logger LOG =
+      org.slf4j.LoggerFactory.getLogger(AlluxioLog4jSocketNode.class);
+  private final AlluxioLogServerProcess mLogServerProcess;
+  private final Socket mSocket;
+
+  /**
+   * Constructor of {@link AlluxioLog4jSocketNode}.
+   *
+   * Callers construct AlluxioLog4jSocketNode instances, passing the ownership of the socket
+   * parameter. From now on, the AlluxioLog4jSocketNode is responsible for closing the socket.
+   *
+   * @param process main log server process
+   * @param socket client socket from which to read {@link org.apache.log4j.spi.LoggingEvent}
+   */
+  public AlluxioLog4jSocketNode(AlluxioLogServerProcess process, Socket socket) {
+    mLogServerProcess = Preconditions.checkNotNull(process,
+        "The log server process cannot be null.");
+    mSocket = Preconditions.checkNotNull(socket, "Client socket cannot be null");
+  }
+
+  @Override
+  public void run() {
+    LoggerRepository hierarchy = null;
+    LoggingEvent event;
+    Logger remoteLogger;
+    try (ObjectInputStream objectInputStream = new ObjectInputStream(
+        new BufferedInputStream(mSocket.getInputStream()))) {
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          event = (LoggingEvent) objectInputStream.readObject();
+        } catch (ClassNotFoundException e1) {
+          throw new RuntimeException("Class of serialized object cannot be found.", e1);
+        } catch (IOException e1) {
+          throw new RuntimeException("Cannot read object from stream due to I/O error.", e1);
+        }
+        if (hierarchy == null) {
+          hierarchy = configureHierarchy(
+              event.getMDC(AlluxioRemoteLogFilter.REMOTE_LOG_MDC_APPENDER_NAME_KEY).toString());
+        }
+        remoteLogger = hierarchy.getLogger(event.getLoggerName());
+        if (event.getLevel().isGreaterOrEqual(remoteLogger.getEffectiveLevel())) {
+          remoteLogger.callAppenders(event);
+        }
+      }
+    } catch (IOException e) {
+      // Something went wrong, cannot recover.
+      throw new RuntimeException("Cannot open ObjectInputStream due to I/O error.", e);
```

Comment:
this could also mean that `ObjectInputStream#close` failed, or that something inside the body of the `try` block threw an `IOException`. I think we can just throw `RuntimeException(e)`

Reply:
Indeed, `configureHierarchy` can also throw `IOException`.